nixops/state/json_file.py

```diff
@@ -0,0 +1,363 @@
+
+        return self.db.read()["deployments"].keys()
+
+    def get_all_deployments(self):
+        """Return Deployment objects for every deployment in the database."""
+        uuids = self.query_deployments()
+        res = []
+        for uuid in uuids:
+            try:
+                res.append(self.open_deployment(uuid=uuid))
+            except nixops.deployment.UnknownBackend as e:
+                sys.stderr.write("skipping deployment ‘{0}’: {1}\n".format(uuid, str(e)))
+        return res
+
+    def _find_deployment(self, uuid=None):
+        all_deployments = self.db.read()["deployments"]
+        found = []
+
+        # if nothing exists no reason to check for things
+        if not all_deployments:
+            return None
+
+        if not uuid:
+            found = filter(lambda(id): all_deployments[id]["attributes"].get("name"), all_deployments)
+        else:
+            found = filter(lambda(id): id == uuid, all_deployments)
+            if not found:
+                found = filter(lambda(id): all_deployments[id]["attributes"].get("name") == uuid, all_deployments)
+            if not found:
+                found = filter(lambda(id): id.startswith(uuid), all_deployments)
+
+        if not found:
+            return None
+
+        if len(found) > 1:
+            if uuid:
+                raise Exception("state file contains multiple deployments with the same name, so you should specify one using its UUID")
+            else:
+                raise Exception("state file contains multiple deployments, so you should specify which one to use using ‘-d’, or set the environment variable NIXOPS_DEPLOYMENT")
+        return nixops.deployment.Deployment(self, found[0], sys.stderr)
+
+    def open_deployment(self, uuid=None):
+        """Open an existing deployment."""
+        deployment = self._find_deployment(uuid=uuid)
+        if deployment: return deployment
+        raise Exception("could not find specified deployment in state file ‘{0}’".format(self.db_file))
+
+    def create_deployment(self, uuid=None):
+        """Create a new deployment."""
+        if not uuid:
+            import uuid
+            uuid = str(uuid.uuid1())
+        with self.db:
+            state = self.db.read()
+            state["deployments"][uuid] = { "attributes": {}, "resources": {} }
+            self.db.set(state)
+        return nixops.deployment.Deployment(self, uuid, sys.stderr)
+
+    def _delete_deployment(self, deployment_uuid):
+        """NOTE: This is UNSAFE, it's guarded in nixops/deployment.py. Do not call this function except from there!"""
+        self.__db.execute("delete from Deployments where uuid = ?", (deployment_uuid,))
+        with self.db:
+            state = self.db.read()
+            state["deployments"].pop(deployment_uuid, None)
+            self.db.set(state)
+
+    def clone_deployment(self, deployment_uuid):
+        with self.db:
+            if not uuid:
```

Comment:
this line was causing errors

```
[nix-shell:~/projects/nixops]$ python2 tests.py tests.functional.test_cloning_clones -vv
nose.config: INFO: Ignoring files matching ['^\\.', '^_', '^setup\\.py$']
nose.config: INFO: Excluding tests matching ['^coverage-tests\\.py$']
nose.selector: INFO: /home/srghma/projects/nixops/dev-shell is executable; skipped
nose.selector: INFO: /home/srghma/projects/nixops/coverage-tests.py is executable; skipped
nose.config: INFO: Ignoring files matching ['^\\.', '^_', '^setup\\.py$']
nose.config: INFO: Excluding tests matching ['^coverage-tests\\.py$']
tests.functional.test_cloning_clones.test_cloning_clones('json', ('ec2', ['/home/srghma/projects/nixops/tests/functional/shared/nix_expressions/logical_base.nix', '/home/srghma/projects/nixops/tests/functional/shared/nix_expressions/ec2_base.nix'])) ... deployment ‘7b0585d1-7144-11e8-95a9-02421304dbb8’ destroyed
ERROR

======================================================================
ERROR: tests.functional.test_cloning_clones.test_cloning_clones('json', ('ec2', ['/home/srghma/projects/nixops/tests/functional/shared/nix_expressions/logical_base.nix', '/home/srghma/projects/nixops/tests/functional/shared/nix_expressions/ec2_base.nix']))
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/nix/store/qyqsfqjpksrvaa387yk4y5bgxnslfyx8-python2.7-nose-1.3.7/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/nix/store/kwq9a1bjhc3w1672khipz85qrwpxj5lv-python2.7-parameterized-0.6.1/lib/python2.7/site-packages/parameterized/parameterized.py", line 280, in <lambda>
    nose_func = wraps(func)(lambda *args: func(*args))
  File "/home/srghma/projects/nixops/tests/functional/test_cloning_clones/__init__.py", line 63, in test_cloning_clones
    deployment2 = deployment.clone()
  File "/home/srghma/projects/nixops/nixops/deployment.py", line 164, in clone
    return self._state.clone_deployment(self.uuid)
  File "/home/srghma/projects/nixops/nixops/state/json_file.py", line 208, in clone_deployment
    if not uuid:
UnboundLocalError: local variable 'uuid' referenced before assignment

----------------------------------------------------------------------
Ran 1 test in 0.098s

FAILED (errors=1)
```


fixed with https://github.com/NixOS/nixops/pull/969/commits/a681561f1b791fa2e791fcccf26a4db016c3aa5a